cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,451 @@
+    using arg_index_input_iterator_t =
+      ArgIndexInputIterator<offset_iterator<InputIteratorT, constant_offset_it_t>, PerPartitionOffsetT, InitT>;
+
+    // The type used for the aggregate that the user wants to find the extremum for
+    using output_aggregate_t = InitT;
+
+    // The output tuple type (i.e., extremum plus index tuples)
+    using per_partition_accum_t = KeyValuePair<PerPartitionOffsetT, output_aggregate_t>;
+    using global_accum_t        = KeyValuePair<GlobalOffsetT, output_aggregate_t>;
+
+    // Unary promotion operator type that is used to transform a per-partition result to a global result
+    // operator()(per_partition_accum_t) -> global_accum_t
+    using promote_to_global_op_t = promote_to_global_op<GlobalOffsetT>;
+
+    // Reduction operator type that enables accumulating per-partition results to a global reduction result
+    using accumulating_transform_output_op_t =
+      accumulating_transform_output_op<global_accum_t, promote_to_global_op_t, ReductionOpT, OutputIteratorT>;
+
+    // The output iterator that implements the logic to accumulate per-partition result to a global aggregate and,
+    // eventually, write to the user-provided output iterators
+    using accumulating_transform_out_it_t =
+      THRUST_NS_QUALIFIER::tabulate_output_iterator<accumulating_transform_output_op_t>;
+
+    // Empty problem initialization type
+    using empty_problem_init_t = empty_problem_init_t<per_partition_accum_t>;
+
+    // Per-partition DispatchReduce template specialization
+    using dispatch_reduce_t =
+      DispatchReduce<arg_index_input_iterator_t,
+                     accumulating_transform_out_it_t,
+                     PerPartitionOffsetT,
+                     ReductionOpT,
+                     empty_problem_init_t,
+                     per_partition_accum_t,
+                     PolicyChainT>;
+
+    void* allocations[2]       = {nullptr};
+    size_t allocation_sizes[2] = {0, 2 * sizeof(global_accum_t)};
+
+    // The current partition's input iterator is an ArgIndex iterator that generates indexes relative to the beginning
+    // of the current partition, i.e., [0, partition_size) along with an OffsetIterator that offsets the user-provided
+    // input iterator by the current partition's offset
+    arg_index_input_iterator_t d_indexed_offset_in(make_offset_iterator(d_in, constant_offset_it_t{GlobalOffsetT{0}}));
+
+    // Transforms the per-partition result to a global result by adding the current partition's offset to the arg result
+    // of a partition
+    promote_to_global_op_t promote_to_global_op{GlobalOffsetT{0}};
+
+    accumulating_transform_output_op_t accumulating_out_op(
+      nullptr, nullptr, false, d_result_out, promote_to_global_op, reduce_op);
+
+    // Upper bound at which we want to cut the input into multiple partitions
+    static constexpr PerPartitionOffsetT max_partition_size = ::cuda::std::numeric_limits<PerPartitionOffsetT>::max();
```

Comment:
So this value will be 2^31-1? Assuming `cub::Reduce` cannot handle exactly 2^31 elements (because the number of elements is no longer representable as `int`), would it make sense to set the maximum to something like `2^32 - 128`? This way, the data for each partition would start at a new cache line, sector, etc. Maybe even a full page makes sense: `2^31-4096`. I don't know how `cub::Reduce` is implemented, but if it would use some vectorized reading this could help.

Reply:
That's a great idea. I'll align the partition size with a page size.